### ISSUE TITLE
Allow formatting of Time using __format__ or to_string

### ISSUE THREAD

**Issue by mhvk** [astropy/astropy#9421]
This would partially address also formatting in Table - see #7902.  The big question for `__format__` is how to deal with a combination of multiple possible formats, both string and numeric, the meaning of precision, etc. Also, should it work only for scalars, or also for arrays? If the latter, how are arrays type-set? Possibly, it may be easiest to start with a `to_string()` method (in analogy with `Angle`).

**Comment by aarchiba** [comment]
I haven't looked into this in any kind of detail yet, but as I understand it Time's civil string subformats already support strptime()-style format specifiers. So using arguments to to_string or format specifier stuff to specify format, subformat, and precision should cover quite a lot - and can probably be extended if necessary by extending format/subformat/precision support?

**Comment by mhvk** [comment]
@aarchiba - agreed, it is not difficult in principle, and as a `to_string()` method even relatively simple. The main trickiness would be for the format specifier in `__format__` - to what extend (if at all) are we going to extend the string formatting mini-language.

**Comment by aarchiba** [comment]
Is it really a question of extending the minilanguage? I mean functionally once python identifies that it is formatting a Time, it sends us everything after the : and we are explicitly free to do what we like with it:

> Objects are able to define their own format specifiers to replace the standard ones. An example is the 'datetime' class, whose format specifiers might look something like the arguments to the strftime() function:

> "Today is: {0:%a %b %d %H:%M:%S %Y}".format(datetime.now())

From PEP 3101 https://www.python.org/dev/peps/pep-3101/#controlling-formatting-on-a-per-type-basis

**Comment by aarchiba** [comment]
So as a first pass, I would be inclined to make to_string accept a format and subformat and maybe a precision (and call str on the result if it's not already a string). Then __format__ would just need to parse its argument into format, subformat, and precision.

How arrays should be handled is another question, as you say, but even just allowing single Times to be formatted would be a big step. And there should be some stuff somewhere about numpy array formatting.

**Comment by mhvk** [comment]
I agree with the first step being `to_string()` - we can do the match later. Of course, `to_string()` can and probably should return string arrays.

**Comment by aarchiba** [comment]
I cast an eye over the code - one complexity is that there is already strftime, which would have some overlap with to_string but only covers "civil" time formats. Coming up with a comfortably orthogonal interface for strftime, to_string, and `__format__` certainly doesn't seem possible before the feature freeze.

**Comment by mhvk** [comment]
Agreed, this is a longer-term thing. It also needs a bit of thought of how to make this consistent with other `to_string()` methods (I just remembered that `SkyCoord` has one too).
